### PR TITLE
Add series annotation and PyOD adapter

### PR DESCRIPTION
**Pull request by fkiraly**
Update: this is now split into two PR, this one for `BaseSeriesAnnotator` (renamed from `BaseStremAnnotator` and one for `BasePanelAnnotator` (PR #1036).

This also includes the `pyod` adapter from #1022, and hence introduces a `pyod>=0.8.0.` soft dependency.

---

This is an extension of @satya-pattnaik's excellent annotation base template #838 that should allow sub-classing or other treatment of non-point-annotation and multi-sample/panel annotation use cases as not uncommon in neuroscience and beyond.

I've created two base classes:

* `BaseStreamAnnotator` which follows most closely @satya-pattnaik's original design
* `BasePanelAnnotator` which extends the design to panel annotation tasks

I've thought long after the first proposal (of mine, i.e., the first commits below) and I think it's not a good idea I had after all, to "merge" the characteristics of both into one. This becomes most clear when you look at what the `update` mode should be - temporal, or adding more sequences?

Changes in both classes:
* rename of series to be annotated from `Z` to `X` - the user expectation is `predict(X)` not `predict(Z)`
* `X` is by default a `pd.DataFrame`, not `pd.Series`. Annotation is typically multivariate, and this is in compliance with the target interface for forecasting.
* adopted nested `fit/_fit` architecture from `forecasting` refactor - it's consensus we're moving towards this, see also #993
* additional, optional argument in `fit`: `Y`, ground truth annotations - for supervised annotation
* allow annotations to be of any `pandas` `dtype`, this allows complex annotations like BIDS (see sktime neuro https://github.com/sktime/sktime-neuro/issues/7)

Additions in the `BaseStreamAnnotator`:

* `update` and `update_predict` functionality, so we can use it in time series streams and directly with forecasters

Additions in `BasePanelAnnotator`:
* `X` can be a list of sequences to be annotated (all `pd.DataFrame`)
* additional, optional argument in `fit` and `predict`: `Z`, `pd.DataFrame` of same `len` as `X` - any metadata on the sequences to annotate.

### CODE REVIEW

**Comment by fkiraly** [comment]
pinging @satya-pattnaik

**Comment by fkiraly** [comment]
@mloning, @satya-pattnaik, I've made a major change since my first commit.

I've meditated about this, and I think we should have two annotator base types - one for panels, one for streams. This splits most naturally by the update mode and use of the argument `Z` (yes/no)

File: sktime/annotation/base/_base.py
Comment (mloning) [inline sktime/annotation/base/_base.py:70]: Let's add tags in a separate PR once we have agreed on them.

File: sktime/annotation/base/_base.py
Comment (mloning) [inline sktime/annotation/base/_base.py:80]: Doesn't this happen in the `BaseEstimator`?

File: sktime/annotation/base/_base.py
Comment (mloning) [inline sktime/annotation/base/_base.py:84]: Based on @Lovkush-A comment, I suggest:
```suggestion
    def fit(self, Z, Y=None, X=None):
```

**Review by mloning — COMMENTED** [review]
I generally agree with the interface refactoring. I also prefer the separation into different kinds of annotators based on input data (e.g. panel and streaming).  

Here are a few suggestions:
* move panel annotators to separate PR, focus this PR on the series annotators, I believe the panel annotation warrants a bit more discussion, ideally in a separate PR. 
* rename `StreamingAnnotator` to `SeriesAnnotator`, this corresponds to the input type and is contrasted with `PanelAnnotator`, also note that panel data may be streaming data. 
* defer implementation of tags until we have agreed on them and can test them. 

I'm happy to merge this quickly, even if we keep a few things open in this PR (e.g. type of annotations, sparse vs dense) and then iterate over it.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: I disagree - I think this should be close to:

* the supervised setting as in `scikit-learn` - we use `fit` and `predict`
* the endogeneous/exogeneous labelling, where information also flows `X`->`Y`

Comment (fkiraly) [reply sktime/annotation/base/_base.py]: also, I think you annotated the wrong annotator? This is the panel annotator, which moved to #1036

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: ok, I commented them out for now

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: based on the suggestion in #1023, I introduced a `transform` method which asliases `predict`.

The `transform` method is compliant with the interface in our transformers, while `predict` mirrors forecasting and `scikit-learn`.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: yes, but no harm done? It's the "safe" thing to do, also looking at the impending `BaseObject` refactor here #891.

File: sktime/annotation/adapters/_pyod.py
Comment (mloning) [inline sktime/annotation/adapters/_pyod.py:10]: Let's follow the naming conventions that we use for transformers already (estimator scitype at the end, ignoring subtypes)
```suggestion
class PyODAnnotator(BaseSeriesAnnotator):
```

File: sktime/annotation/adapters/_pyod.py
Comment (mloning) [inline sktime/annotation/adapters/_pyod.py:39]: * Perhaps also worth shorting them to `format` and `values` since the annotation context is clear. 
* Should `values` be renamed to `labels`? Values may be a bit generic, labels will be familiar from the classification setting. 
* I suggest to write helper functions for input checks, e.g. `_check_format(format)`, and add them to `sktime/utils/validation/annotation.py`

File: sktime/annotation/adapters/_pyod.py
Comment (mloning) [inline sktime/annotation/adapters/_pyod.py:75]: Since estimator is technically a hyper-parameter we don't want to overwrite it here (none of the methods should overwrite hyper-parameters except the constructor and `set_params`), so instead we usually do:
```suggestion
        self.estimator_ = clone(self.estimator)
        self.estimator_.fit(X_np)
```
with trailing underscores indicating fitted attributes.

File: sktime/annotation/base/_base.py
Comment (mloning) [inline sktime/annotation/base/_base.py:179]: If we want to treat this as a transformer, should it not also inherit from one of the transformer types? It may be easier to implement this as a wrapper.

File: sktime/annotation/base/_mock.py
Comment (mloning) [inline sktime/annotation/base/_mock.py:5]: Since we have the `PyODAnnotator` in this PR, I think we can remove the mock class

File: sktime/annotation/tests/test_all_annotators.py
Comment (mloning) [inline sktime/annotation/tests/test_all_annotators.py:1]: can we add more tests on the output, i.e. if format=dense, it should have the same index as X

File: sktime/utils/_testing/estimator_checks.py
Comment (mloning) [inline sktime/utils/_testing/estimator_checks.py:135]: Are the changes to the tag system necessary for this PR already?

**Review by mloning — COMMENTED** [review]
Thanks @fkiraly and @satya-pattnaik looks very good already, I left a few comments below, should be quick to fix, let me know if you have any questions!

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py:179]: what would we get from inheriting?

File: sktime/annotation/adapters/_pyod.py
Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:75]: do you know whether this works with `pyod` models out-of-the-box?

File: sktime/annotation/adapters/_pyod.py
Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:39]: >since the annotation context is clear.

yes, but the context that this is something that applies to the output is not clear.
How about `y_format` etc?

Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:39]: >I suggest to write helper functions for input checks,

With this design, would that not be identical with `check_series`?

Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:39]: >Values may be a bit generic, labels will be familiar from the classification setting.

Yes, good idea

File: sktime/annotation/adapters/_pyod.py
Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:10]: sure - generally I'd prefer the scitypes at the start, since then the lexicographic ordering clusters scitypes, but obviously this is the wrong place to change that. I'll comply with current standard.

**Comment by fkiraly** [comment]
Any good idea where this is coming from?
` ValueError: Invalid estimator type: <class 'sktime.annotation.base._mock.MockSeriesAnnotator'>`

I suspect it's coming from incorrect registration of the estimator in some test config, but can't track it down easily.

File: sktime/annotation/tests/test_all_annotators.py
Comment (fkiraly) [reply sktime/annotation/tests/test_all_annotators.py:1]: hm, whould we move the `format` arguments into the base class and pass them via `super`? That way it would be an expected interfacve point there, and can be read in the base class for checks.

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py:179]: maybe, but do we get anything from the wrapper that isn't already addressed by this interface alias?

File: sktime/utils/_testing/estimator_checks.py
Comment (fkiraly) [reply sktime/utils/_testing/estimator_checks.py:135]: no, but to revert this I would have to go through multiple files at the cost of time and the risk of breaking things - it was already a hassle to split the PR into two

File: sktime/annotation/adapters/_pyod.py
Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:39]: it already did `check_series` when you wrote the above, so can you please explain what you would want on top of that?

Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py:39]: >should values be renamed to labels

there actually already is a variable that is called "labels", it's the one which tells you the type of labels

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: this was inconsistent with the `fit`, so I added aliasing (of `X` and `Z`) to all `fit`-like methods

File: sktime/annotation/adapters/_pyod.py
Comment (mloning) [reply sktime/annotation/adapters/_pyod.py]: I would think so, they're closely aligned with scikit-learn, but worth checking.

Comment (fkiraly) [reply sktime/annotation/adapters/_pyod.py]: seems to work - good

File: sktime/annotation/base/_base.py
Comment (fkiraly) [reply sktime/annotation/base/_base.py]: I actually removed that agin. I think now it was a bad idea.

We may want to have some aliasing or not, but first we should refactor the transformers modules.
I decided this is out of scope for this PR, probably you agree.

File: sktime/annotation/base/_mock.py
Comment (fkiraly) [reply sktime/annotation/base/_mock.py:5]: will do once the `PyODAnnotator` works

**Review by satya-pattnaik — COMMENTED** [review]
Really nice work @fkiraly . Will need to add more unit tests to it. Let me know if you have some in your mind. 
Also to remove the mock annotator.

File: sktime/annotation/tests/test_all_annotators.py
Comment (mloning) [reply sktime/annotation/tests/test_all_annotators.py:1]: Yes, moving them to `super` sounds good to me

File: sktime/annotation/adapters/_pyod.py
Comment (mloning) [reply sktime/annotation/adapters/_pyod.py:41]: For readability, I'd write:
```python
def check_fmt(fmt):
    valid_fmts = ["sparse", "dense"]
    if fmt not in valid_fmts:
        raise ValueError(f"`fm` must be in: {valid_fmts}, but found: {fmt}.")
    return fmt
```
We shouldn't use `format`, that's already a Python function.

**Review by mloning — COMMENTED** [review]
Hi @fkiraly, let me know if you need any help with this PR, I wanted to work on series annotation anyway.

**Comment by fkiraly** [comment]
>let me know if you need any help with this PR, I wanted to work on series annotation anyway.

sure - I think there's not much outstanding:

* tests, potentially writing some specifically for annotators
* moving the sparse/dense setting to the base class constructor, and passing the settings from concrete classes to the base class constructor
* updating the extension template with any additions

**Comment by mloning** [comment]
Okay, I'll take a look at this now

**Comment by mloning** [comment]
Please see #1093 for additional changes, this should address the open conversations on this PR. 

Once that is merged, we can merge #1064.

Quick questions: You can only set "sparse" with "indicator", right, or does "sparse" also work with "score"?

**Review by mloning — APPROVED** [review]
I think this is good to go, but would wait after the release, any other comments?

**Comment by fkiraly** [comment]
>but would wait for after the release

Why that?

Doesn't hurt to have another base class and extension templates in.
That way, the community can more easily help extending the annotation module.

**Comment by mloning** [comment]
@fkiraly I just though we would want to get the basic unit testing done first before we publish it. 

Are you happy with this being merged as it is?

**Comment by mloning** [comment]
Thinking about it, perhaps it would be nice to get this out with the new release? Feel free to merge this @fkiraly.

**Comment by fkiraly** [comment]
>Thinking about it, perhaps it would be nice to get this out with the new release? Feel free to merge this @fkiraly.

Yes - might be a good template for people to contribute to. The individual contributions may then lead to base class refactor, which may lead to a major version up. So, agreed.

**Comment by mloning** [comment]
Now merged @satya-pattnaik @fkiraly - great work, it took us some time, but we finally have some basic annotation framework set up!  🎉

**Pull request merged**
